operator/pkg/certs/certs.go

```diff
@@ -0,0 +1,456 @@
+		NotBefore:             caCert.NotBefore,
+		NotAfter:              notAfter,
+		KeyUsage:              keyUsage,
+		ExtKeyUsage:           cc.Config.Usages,
+		BasicConstraintsValid: true,
+		IsCA:                  isCA,
+	}
+	certDERBytes, err := x509.CreateCertificate(cryptorand.Reader, &certTmpl, caCert, key.Public(), caKey)
+	if err != nil {
+		return nil, err
+	}
+	return x509.ParseCertificate(certDERBytes)
+}
+
+// RemoveDuplicateAltNames removes duplicate items in altNames.
+func RemoveDuplicateAltNames(altNames *certutil.AltNames) {
+	if altNames == nil {
+		return
+	}
+
+	if altNames.DNSNames != nil {
+		altNames.DNSNames = sets.NewString(altNames.DNSNames...).List()
+	}
+
+	ipsKeys := make(map[string]struct{})
+	var ips []net.IP
+	for _, one := range altNames.IPs {
+		if _, ok := ipsKeys[one.String()]; !ok {
+			ipsKeys[one.String()] = struct{}{}
+			ips = append(ips, one)
+		}
+	}
+	altNames.IPs = ips
+}
+
+func appendSANsToAltNames(altNames *certutil.AltNames, SANs []string) {
+	for _, altname := range SANs {
+		if ip := netutils.ParseIPSloppy(altname); ip != nil {
+			altNames.IPs = append(altNames.IPs, ip)
+		} else if len(validation.IsDNS1123Subdomain(altname)) == 0 {
+			altNames.DNSNames = append(altNames.DNSNames, altname)
+		} else if len(validation.IsWildcardDNS1123Subdomain(altname)) == 0 {
+			altNames.DNSNames = append(altNames.DNSNames, altname)
+		}
+	}
+}
+
+// EncodeCertPEM returns PEM-endcoded certificate data
+func EncodeCertPEM(cert *x509.Certificate) []byte {
+	block := pem.Block{
+		Type:  CertificateBlockType,
+		Bytes: cert.Raw,
+	}
+	return pem.EncodeToMemory(&block)
+}
+
+// ParsePrivateKeyPEM parses crypto.Signer from byte array. the key
+// must be encryption by ECDSA and RAS.
+func ParsePrivateKeyPEM(keyData []byte) (crypto.Signer, error) {
+	caPrivateKey, err := keyutil.ParsePrivateKeyPEM(keyData)
+	if err != nil {
+		return nil, err
+	}
+
+	// Allow RSA and ECDSA formats only
+	var key crypto.Signer
+	switch k := caPrivateKey.(type) {
+	case *rsa.PrivateKey:
+		key = k
+	case *ecdsa.PrivateKey:
+		key = k
+	default:
+		return nil, errors.New("the private key is neither in RSA nor ECDSA format")
+	}
+
+	return key, nil
+}
+
+func makeAltNamesMutator(f func(cfg *AltNamesMutatorConfig) (*certutil.AltNames, error)) altNamesMutatorFunc {
+	return func(cfg *AltNamesMutatorConfig, cc *CertConfig) error {
+		altNames, err := f(cfg)
+		if err != nil {
+			return err
+		}
+
+		cc.Config.AltNames = *altNames
+		return nil
+	}
+}
+
+func etcdServerAltNamesMutator(cfg *AltNamesMutatorConfig) (*certutil.AltNames, error) {
+	etcdClientServiceDNS := fmt.Sprintf("%s.%s.svc.cluster.local", util.KarmadaEtcdClientName(cfg.Name), cfg.Namespace)
+	etcdPeerServiceDNS := fmt.Sprintf("*.%s.%s.svc.cluster.local", util.KarmadaEtcdName(cfg.Name), cfg.Namespace)
+
+	altNames := &certutil.AltNames{
+		DNSNames: []string{"localhost", etcdClientServiceDNS, etcdPeerServiceDNS},
+		IPs:      []net.IP{net.IPv4(127, 0, 0, 1)},
```

Comment:
Could we add host node ip here automatically, or make it configurable? 

Because we can export `karmada-apiserver.config` from secret, but it was not valid for my node ip when I made the `karmada-apiserver sverive` to `NodePort` type, it only support `127.0.0.1`

```
[root@master67 config]# kubectl get clusters --kubeconfig karmada-apiserver.conf
Unable to connect to the server: x509: certificate is valid for 127.0.0.1, not 10.10.103.67
```

It is necessary for joining a PULL mod Cluster.

Reply:
Ok, I got it. If we want to expose the svc of karmada apiserver, we must add some `alts` to certificates.
But, There are several mode here: `ClusterIP`, `NodePort` or `LB`, different mode have different alts. I haven't figured out how to implement the feature. So, I plan to implement the feature later on with this pr.